GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -55,28 +56,37 @@ public void Dispose()
         private void TimerCallback(object unusedState)
         {
             string errorMessage = null;
-
             InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
-            ProductUpgrader productUpgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
-            if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
+            IProductUpgrader productUpgrader;
+
+            if (ProductUpgrader.TryCreateUpgrader(out productUpgrader, this.tracer, out errorMessage))
             {
-                return;
-            }
+                if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
+                {
+                    return;
+                }
 
-            productUpgrader.CleanupDownloadDirectory();
+                productUpgrader.CleanupDownloadDirectory();
+            }
 
             if (errorMessage != null)
             {
                 this.tracer.RelatedError(errorMessage);
             }
         }
 
-        private bool TryDownloadUpgrade(ProductUpgrader productUpgrader, out string errorMessage)
+        private bool TryDownloadUpgrade(IProductUpgrader productUpgrader, out string errorMessage)
         {
             using (ITracer activity = this.tracer.StartActivity("Checking for product upgrades.", EventLevel.Informational))
             {
                 Version newerVersion = null;
                 string detailedError = null;
+                bool isError;
+                if (!productUpgrader.CanRunUsingCurrentConfig(out isError, out errorMessage))

```

Comment:
Why did we not need to check this before?

Reply:
The code was there before in `UpgradeVerb` & `UpgradeOrchestrator`, but not as a separate method. Now with upcoming Nuget upgrader, it made sense to make it as a separate interface method, so the verb and orchestrator don't need to know about Github release rings or Nuget feed URLs.